jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/EventsHandler.java

```diff
@@ -0,0 +1,419 @@
+ * <p>It is recommended that the event parameters are <em>not</em> acted upon
+ * in the listener methods, or undefined behavior may result. On the other
+ * hand, it is legit to store request attributes in one listener method that
+ * may be possibly retrieved in another listener method in a later event.</p>
+ * <p>Listener methods are invoked synchronously from the thread that is
+ * performing the request processing, and they should not call blocking code
+ * (otherwise the request processing will be blocked as well).</p>
+ * <p>The kind of chunk passed to {@link #onRequestRead(Request, Content.Chunk)} depends on
+ * the parent of this handler. For instance, if the parent is the Server, then raw chunks
+ * are always passed. If somewhere in the parent chain is the {@code GzipHandler} then
+ * unzipped chunks are passed.</p>
+ */
+public abstract class EventsHandler extends Handler.Wrapper
+{
+    private static final Logger LOG = LoggerFactory.getLogger(EventsHandler.class);
+    
+    public EventsHandler()
+    {
+    }
+
+    public EventsHandler(Handler handler)
+    {
+        super(handler);
+    }
+
+    @Override
+    public boolean handle(Request request, Response response, Callback callback) throws Exception
+    {
+        // Before handling
+        ReadOnlyRequest roRequest = new ReadOnlyRequest(request);
+        fireOnBeforeHandling(roRequest);
+        try
+        {
+            EventsRequest wrappedRequest = new EventsRequest(request, roRequest);
+            EventsResponse wrappedResponse = new EventsResponse(roRequest, response);
+            request.addHttpStreamWrapper(stream -> new HttpStream.Wrapper(stream)
+            {
+                private void fireNotifications(Throwable failure)
+                {
+                    if (!wrappedResponse.notifiedOnResponseBegin)
+                        fireOnResponseBegin(roRequest, wrappedResponse.getHeaders().asImmutable(), wrappedResponse.getStatus());
+                    if (wrappedResponse.suppliedTrailers != null)
+                        fireOnResponseTrailersComplete(roRequest, wrappedResponse.suppliedTrailers);
+                    fireOnComplete(roRequest, failure);
+                }
+
+                @Override
+                public void succeeded()
+                {
+                    fireNotifications(null);
+                    super.succeeded();
+                }
+
+                @Override
+                public void failed(Throwable x)
+                {
+                    fireNotifications(x);
+                    super.failed(x);
+                }
+            });
+
+            boolean handled = super.handle(wrappedRequest, wrappedResponse, callback);
+
+            fireOnAfterHandling(roRequest, handled, null);
+            return handled;
+        }
+        catch (Throwable x)
+        {
+            fireOnAfterHandling(roRequest, false, x);
+            throw x;
+        }
+    }
+
+    private void fireOnBeforeHandling(Request request)
```

Comment:
I don't mind the "fireOn" names, but elsewhere we have used "notify" for the same purpose, so I'm +0 for changing that... but only if you really want to.

Reply:
`notify` it shall be.